platform/android/src/jni.cpp

```diff
@@ -1620,6 +1753,363 @@ void JNICALL nativeRemoveCustomLayer(JNIEnv *env, jobject obj, jlong nativeMapVi
+}
+
+jstring JNICALL getAccessToken(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "getAccessToken");
+    assert(defaultFileSourcePtr != 0);
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    std::string accessToken = defaultFileSource->getAccessToken();
+    return std_string_to_jstring(env, accessToken);
+}
+
+void JNICALL listOfflineRegions(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr, jobject listCallback) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "listOfflineRegions");
+
+    // Checks
+    assert(defaultFileSourcePtr != 0);
+    if (listCallback == nullptr) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Callback has to be set.");
+        return;
+    }
+
+    // Makes sure the callback doesn't get GC'ed
+    listCallback = reinterpret_cast<jobject>(env->NewGlobalRef(listCallback));
+
+    // Launch listCallback
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    defaultFileSource->listOfflineRegions([defaultFileSourcePtr, listCallback](std::exception_ptr error, mbgl::optional<std::vector<mbgl::OfflineRegion>> regions) {
+
+        // Reattach, the callback comes from a different thread
+        JNIEnv *env2;
+        jboolean renderDetach = attach_jni_thread(theJVM, &env2, "Offline Thread");
+        if (renderDetach) {
+            mbgl::Log::Debug(mbgl::Event::JNI, "Attached.");
+        }
+
+        if (error) {
+            std::string message = mbgl::util::toString(error);
+            env2->CallVoidMethod(listCallback, listOnErrorMethodId, std_string_to_jstring(env2, message));
+        } else if (regions) {
+            // Build the OfflineManager object
+            jobject jmanager = env2->NewObject(offlineManagerClass, offlineManagerClassConstructorId);
+            env2->SetLongField(jmanager, offlineManagerClassPtrId, defaultFileSourcePtr);
+
+            // Build jobjectArray
+            jsize index = 0;
+            jobjectArray jregions = env2->NewObjectArray(regions->size(), offlineRegionClass, NULL);
+            for (std::vector<mbgl::OfflineRegion>::iterator region = regions->begin(); region != regions->end(); ++region) {
+                // Build the Region object
+                jobject jregion = env2->NewObject(offlineRegionClass, offlineRegionConstructorId);
+                env2->SetObjectField(jregion, offlineRegionOfflineManagerId, jmanager);
+                env2->SetLongField(jregion, offlineRegionIdId, region->getID());
+                env2->SetObjectField(jregion, offlineRegionMetadataId, metadata_from_native(env2, region->getMetadata()));
+
+                // Moves the region on the stack into a heap-allocated one
+                env2->SetLongField(jregion, offlineRegionPtrId,
+                    reinterpret_cast<jlong>(new mbgl::OfflineRegion(std::move(*region))));
```

Comment:
No, the Java `OfflineRegion`. Wherever you have a Java instance with ownership of a corresponding C++ instance, as you do here with `OfflineRegion` and `mbgl::OfflineRegion`, the Java instance needs to release the C++ instance in its finalizer.

The other place this applies is `OfflineManager`, which owns a native reference to a `mbgl::DefaultFileSource`. It's less of a concern there though since `OfflineManager` is a singleton.